### PR TITLE
[Yaml] Recommend using quotes instead of PARSE_KEYS_AS_STRINGS

### PR DESCRIPTION
**Pull request by GuilhemN**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no
| Deprecations? | yes 
| Tests pass?   | yes
| Fixed tickets | #... 
| License       | MIT
| Doc PR        | 

Sorry for opening this so lately... I just realized that we could get rid of `Yaml::PARSE_KEYS_AS_STRINGS` just by recommending using quotes...

~This way we don't allow a behavior not respecting the spec and we won't need to deprecate `PARSE_KEYS_AS_STRINGS` later.~

~Is it too late for this to be merged in 3.3?~

ping @xabbuh

### CODE REVIEW

File: src/Symfony/Component/Yaml/Inline.php [hunk 1]
```diff
@@ -490,11 +490,10 @@ private static function parseMapping($mapping, $flags, &$i = 0, $references = ar
                 @trigger_error('Omitting the key of a mapping is deprecated and will throw a ParseException in 4.0.', E_USER_DEPRECATED);
             }
 
-            if (!(Yaml::PARSE_KEYS_AS_STRINGS & $flags)) {
+            if (!$isKeyQuoted) {
```
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Inline.php]: Added because quoted keys should not be evaluated

File: src/Symfony/Component/Yaml/Inline.php [hunk 2]
```diff
@@ -490,11 +490,10 @@ private static function parseMapping($mapping, $flags, &$i = 0, $references = ar
...
                 $evaluatedKey = self::evaluateScalar($key, $flags, $references);
-
-                if ('' !== $key && $evaluatedKey !== $key && !is_string($evaluatedKey)) {
-                    @trigger_error('Implicit casting of incompatible mapping keys to strings is deprecated since version 3.3 and will throw \Symfony\Component\Yaml\Exception\ParseException in 4.0. Pass the PARSE_KEYS_AS_STRING flag to explicitly enable the type casts.', E_USER_DEPRECATED);
+                if ('' !== $key && $evaluatedKey !== $key && !is_string($evaluatedKey) && !is_int($evaluatedKey)) {
```
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Inline.php]: I added the `is_int` check because unquoted integers work fine

**Comment by xabbuh** [comment]
I do not think this really is an option in general. The reason is that you may not be able to influence how the YAML file is actually dumped (e.g. an external data source) and not having quotes around mapping keys is totally valid.

**Comment by GuilhemN** [comment]
Using `PARSE_KEYS_AS_STRINGS` in our loaders mean that they don't evaluate keys and so do not comply with the spec.
At the same time, I don't think we should be able to parse a yaml file using features not supported by the language because it loses its meaning (if you use false as a key it means something, if you use "false" it means something else).

> The reason is that you may not be able to influence how the YAML file is actually dumped

Let's be pragmatic, I don't think someone will ever use `false`, `true`, or whatever as a mapping key so imo we should be protected against these cases, because php doesn't support them so you can't reflect these values in the result you get from the parse.
[The php extension](https://secure.php.net/manual/fr/book.yaml.php) always evaluate mapping keys but objects are serialized, `false` is converted to `""`, `true` to `1`, `null` to `""`. I think this is wrong because most do not know this and you can't know what was the exact original value.

> not having quotes around mapping keys is totally valid

It shouldn't be for values that are not supported by php imo because you end up with a weird behavior for basically no gain (who would voluntarily ever use these values as keys?).

**Comment by GuilhemN** [comment]
Updated to target 3.4

**Comment by GuilhemN** [comment]
Phpunit fails on 4.0 because of https://github.com/symfony/symfony/pull/22770, [this exception](https://github.com/symfony/symfony/blame/aa4fdc0ec5d795b3fb84b82471ce613d8bec287b/src/Symfony/Component/Yaml/Inline.php#L444) must not be thrown when an integer is parsed (it is supported by php so there's no problem).

**Comment by xabbuh** [comment]
@GuilhemN That's fixed by #23108, right?

**Comment by GuilhemN** [comment]
@xabbuh indeed, pr rebased. The build still fails but this is not caused by this PR.

**Review by fabpot — APPROVED** [review]
@xabbuh Is this one ok for you?

**Comment by nicolas-grekas** [comment]
@GuilhemN minor conflict to resolve

**Comment by nicolas-grekas** [comment]
rebase please, not merge :) (online editing creates merges :( )

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 3]
```diff
@@ -1843,27 +1818,13 @@ public function testPhpConstantTagMappingKey()
         $this->assertSame($expected, $this->parser->parse($yaml, Yaml::PARSE_CONSTANT));
     }
 
-    public function testPhpConstantTagMappingKeyWithKeysCastToStrings()
```
Comment (stof) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: IMO, this test should not disappear. We need to keep it to ensure this case keeps working in 3.4 (it should become a legacy test though)

File: src/Symfony/Component/Yaml/Parser.php [hunk 4]
```diff
@@ -236,9 +240,9 @@ private function doParse($value, $flags)
                     throw $e;
                 }
 
-                if (!(Yaml::PARSE_KEYS_AS_STRINGS & $flags) && !is_string($key) && !is_int($key)) {
+                if (!is_string($key) && !is_int($key)) {
```
Comment (stof) [inline src/Symfony/Component/Yaml/Parser.php]: are you sure about this change ?

File: src/Symfony/Component/Yaml/Parser.php [hunk 5]
```diff
@@ -236,9 +240,9 @@ private function doParse($value, $flags)
...
                     $keyType = is_numeric($key) ? 'numeric key' : 'non-string key';
-                    @trigger_error(sprintf('Implicit casting of %s to string is deprecated since version 3.3 and will throw \Symfony\Component\Yaml\Exception\ParseException in 4.0. Pass the PARSE_KEYS_AS_STRINGS flag to explicitly enable the type casts.', $keyType), E_USER_DEPRECATED);
+                    @trigger_error(sprintf('Implicit casting of %s to string is deprecated since version 3.3 and will throw \Symfony\Component\Yaml\Exception\ParseException in 4.0. Quote your evaluable mapping keys instead.', $keyType), E_USER_DEPRECATED);
```
Comment (stof) [inline src/Symfony/Component/Yaml/Parser.php]: I suggest changing this message in 3.3 already. People who have not yet migrated to 3.3 should get the advice of quoting directly, to avoid a double migration for them. Quoting should already be working in 3.3.

File: src/Symfony/Component/Routing/Loader/YamlFileLoader.php [hunk 6]
```diff
@@ -59,7 +59,7 @@ public function load($file, $type = null)
         }
 
         try {
-            $parsedConfig = $this->yamlParser->parse(file_get_contents($path), Yaml::PARSE_KEYS_AS_STRINGS);
+            $parsedConfig = $this->yamlParser->parse(file_get_contents($path));
```
Comment (stof) [inline src/Symfony/Component/Routing/Loader/YamlFileLoader.php]: isn't this a BC break ?

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Routing/Loader/YamlFileLoader.php (hunk 6 above)
Comment (GuilhemN) [inline src/Symfony/Component/Routing/Loader/YamlFileLoader.php]: Not more than when we added it in 3.3 (for instance `!str foo` in routing files was parsed as `"foo"` in 3.2 and is as `"!str foo"` since 3.3). Reverting this change is in fact more intuitive imo as it does make the parser always spec compliant.

File: src/Symfony/Component/Yaml/Parser.php (hunk 5 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: I agree, I'll submit a pr if this is merged.

File: src/Symfony/Component/Yaml/Parser.php (hunk 4 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: Yes, evaluable keys are no longer allowed unless they're quoted, even when using `Yaml::PARSE_KEYS_AS_STRINGS`.

File: src/Symfony/Component/Yaml/Tests/ParserTest.php (hunk 3 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: I don't remember why i removed it so I guess we can revert this. However I won't be able to do it for the next 2 weeks, feel free to push on my repository if you have the time, otherwise I'll do the change ASAP.

**Comment by xabbuh** [comment]
Recommending to quote keys indeed looks like a good idea. However, I am not convinced we should deprecate the `PARSE_KEYS_AS_STRINGS`. It can still be useful when you try to parse some YAML files that you do not control.

**Comment by GuilhemN** [comment]
@xabbuh imo it causes too many issues for what it's worth. 
For instance `!str true: foo` should work fine but with the flag, you'll end up with a key you don't expect. We have the same problem with other syntax such as `! true` and  `!str "foo"`.
Because of this you also won't be able to detect colliding keys (`1` and `0x1` for example).

If the yaml file you try to parse uses features not supported by the sf parser nor by PHP, it looks normal and expectable to me that you can't parse this file.

Lastly, I don't think this fits something else than very edge cases that could probably also be fixed by altering the yaml file parsed.

File: src/Symfony/Component/Yaml/Tests/ParserTest.php (hunk 3 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: done

File: src/Symfony/Component/Yaml/Parser.php (hunk 5 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: see https://github.com/symfony/symfony/pull/23635

File: src/Symfony/Component/Validator/Mapping/Loader/YamlFileLoader.php [hunk 7]
```diff
@@ -14,7 +14,6 @@
 use Symfony\Component\Validator\Mapping\ClassMetadata;
 use Symfony\Component\Yaml\Exception\ParseException;
 use Symfony\Component\Yaml\Parser as YamlParser;
-use Symfony\Component\Yaml\Yaml;
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Mapping/Loader/YamlFileLoader.php]: this must be reverted

File: UPGRADE-3.4.md [hunk 8]
```diff
@@ -111,3 +111,31 @@ Yaml
...
+   2.0: float key
+   YAML;
+
+   Yaml::parse($yaml, Yaml::PARSE_KEYS_AS_STRINGS);
+   ```
+
+   After:
+
+   ```php
+
```
Comment (xabbuh) [inline UPGRADE-3.4.md]: the blank line can be removed

File: UPGRADE-4.0.md [hunk 9]
```diff
@@ -666,15 +665,41 @@ Yaml
...
    2.0: float key
    YAML;
 
    Yaml::parse($yaml, Yaml::PARSE_KEYS_AS_STRINGS);
    ```
 
+   After:
+
+   ```php
+
```
Comment (xabbuh) [inline UPGRADE-4.0.md]: can be removed

File: src/Symfony/Component/Routing/Loader/YamlFileLoader.php (hunk 6 above)
Comment (xabbuh) [inline src/Symfony/Component/Routing/Loader/YamlFileLoader.php]: Shouldn't we do this change in #23635 then?

File: src/Symfony/Component/Yaml/Tests/InlineTest.php [hunk 10]
```diff
@@ -730,22 +730,14 @@ public function testTheEmptyStringIsAValidMappingKey()
...
      */
     public function testImplicitStringCastingOfMappingKeysIsDeprecated($yaml, $expected)
     {
         $this->assertSame($expected, Inline::parse($yaml));
     }
 
-    /**
-     * @dataProvider getNotPhpCompatibleMappingKeyData
-     */
-    public function testExplicitStringCastingOfMappingKeys($yaml, $expected)
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/InlineTest.php]: IMO the test should be kept but being flagged as legacy

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 11]
```diff
@@ -71,19 +71,6 @@ public function getDataFormSpecifications()
     }
 
     /**
-     * @dataProvider getNonStringMappingKeysData
-     */
-    public function testNonStringMappingKeys($expected, $yaml, $comment)
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: same here

**Review by xabbuh — CHANGES_REQUESTED** [review]

**Comment by xabbuh** [comment]
@GuilhemN I would still expect floats to be used more frequently as keys than tags, but if noone else objects, I am fine with the proposed changes. I just left some minor comments.

**Comment by GuilhemN** [comment]
@xabbuh maybe but people just have to quote the key if they want to use a float while we have no solution for tags. And still the same argument, the yaml file is invalid and could be rejected by another parser :)

File: src/Symfony/Component/Yaml/Tests/InlineTest.php (hunk 10 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Tests/InlineTest.php]: done

File: src/Symfony/Component/Routing/Loader/YamlFileLoader.php (hunk 6 above)
Comment (GuilhemN) [inline src/Symfony/Component/Routing/Loader/YamlFileLoader.php]: I prefer doing this in a minor but if you insist I can of course do it in 3.3.

File: src/Symfony/Component/Serializer/Mapping/Loader/YamlFileLoader.php [hunk 12]
```diff
@@ -114,7 +114,7 @@ private function getClassesFromYaml()
             $this->yamlParser = new Parser();
         }
 
-        $classes = $this->yamlParser->parse(file_get_contents($this->file), Yaml::PARSE_KEYS_AS_STRINGS);
+        $classes = $this->yamlParser->parse(file_get_contents($this->file));
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Mapping/Loader/YamlFileLoader.php]: Can the `use` statement be removed too?

File: src/Symfony/Component/Routing/Loader/YamlFileLoader.php (hunk 6 above)
Comment (xabbuh) [inline src/Symfony/Component/Routing/Loader/YamlFileLoader.php]: The `use` statement can be removed too?

File: src/Symfony/Component/Serializer/Mapping/Loader/YamlFileLoader.php (hunk 12 above)
Comment (GuilhemN) [inline src/Symfony/Component/Serializer/Mapping/Loader/YamlFileLoader.php]: indeed

**Comment by xabbuh** [comment]
@GuilhemN Can you rebase here? #23635 has been merged up.

**Comment by GuilhemN** [comment]
@xabbuh done

**Review by xabbuh — APPROVED** [review]

**Review by nicolas-grekas — COMMENTED** [review]
shouldn't the `@deprecated` annotation be added to PARSE_KEYS_AS_STRINGS in Yaml.php?

**Comment by GuilhemN** [comment]
@nicolas-grekas indeed.

File: src/Symfony/Component/Yaml/Yaml.php [hunk 13]
```diff
@@ -31,6 +31,11 @@ class Yaml
     const PARSE_CONSTANT = 256;
     const PARSE_CUSTOM_TAGS = 512;
     const DUMP_EMPTY_ARRAY_AS_SEQUENCE = 1024;
+
+    /**
+     * @deprecated since version 3.4, to be removed in 4.0. Quote your evaluable
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Yaml.php]: Should be on one line

Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Yaml.php]: Indeed :)

**Comment by xabbuh** [comment]
Thank you @GuilhemN.
